### DISCUSSION TITLE
CraftCMS programmatically copy Link field value to URL field

### DISCUSSION THREAD

**Discussion by snagelia** [craftcms/cms#15032] [Q&A]
Hi,
We are running multisite CraftCMS 4.9.x and heavily using Linked field for many entry types. Now, Craft 5 doesn't support Linked field, so I am planning to convert the existing Linked Field to URL field. Now my problem is we are having so many entries that doing manually will take long time. So thinking to do it programmatically by he following steps:
1. Create a new URL field 
2.  Then copy the existing entry.id.Linkedfield.url to the new URL field

I never edit fields programmatically in Craft so not sure how to do it, where to create this file and how to run it. Any details and steps would be very helpful OR any other better way to do it. 

Thanks in advance.

**Comment by theAdhocracy** [comment]
I've been doing a lot of field migration after rolling out Craft 5. I don't have a definitive answer (others may know of easier/better techniques) but here are a few that have been working well:

- Craft migrations. Check out this article for some good example code and explanations: https://nystudio107.com/blog/adding-content-with-content-migrations-in-craft-cms
- FeedMe (plugin). You can either create a CSV or JSON export of the entry types containing the relevant fields, then create a new import in FeedMe that uses that to populate the new field, or generate an API endpoint (the Element API plugin is super useful here) and point to that. I've done both, they work well for this kind of cloning text between nodes.
- Changing the field type directly. You'd definitely want to test this locally first, but some field types "just work" if you swap between them. A clear example is Redactor <=> CKEditor, but as long as the underlying data is similar, I've seen this work in a few places e.g. Plain Text <=> URL. I haven't used the Linked field myself, but if it's storing a text string then you might find changing to a Plain Text or URL field will preserve the existing values.
- Bulk editing. Craft 5 has a solid implementation baked in, but there's also the Venveo Bulk Edit plugin which works great on Craft 3/4 in my experience. Both are slightly different, but let you edit multiple entries at once. If you're cloning values between fields, combine them with local scripts run via the console in your browser's developer tools. That lets you execute small JavaScript functions, so you can use DOM traversal to find the old field and copy its value into a variable, then traverse to the new field and update its `value` attribute. Do that in a `forEach` loop and you can complete entire batches of entries in a single go. I'd recommend using Filters to ensure you aren't repeating entries (e.g. filter for only entries where the new field is blank) and you can get through a few thousand in half an hour relatively easily. Here's a partially updated example of one I've been using:

```
[].forEach.call(
  document.querySelectorAll("td[data-title='Link'] input[type='text']"), 
  function(e) { 
    const current = e.value;
    const sibling = e.parentElement.nextSibling.querySelector("input[type='url']");
    sibling.value = current;
  }
);
```

With all of the above, take database backups, test them locally first, and rigorously stress test. We've had some issues where unrelated data on entries seems to get lost (not sure why) using the native bulk edit, for instance, but unable to replicate the results, so hard to tell why.

**Reply by snagelia** [reply]
Thank you theAdocracy for your answer. I have though about FeedMe and going to try that. For Bulk editing, can you provide little more details about the process like where to create the file under the CraftCMS project? Will this process update the value in Craft database? Do I run the script for each page in my site? I have many many pages for my website. Sorry for asking dumb question. Thanks again.

**Reply by theAdhocracy** [reply]
For Bulk Editing, you aren't running any files. If you're on Craft 4, install the Bulk Edit plugin from the plugin store. I'd look at their docs directly to see how far you can get, but my guess is that to automate copying the field data you'd need to use some local JavaScript. Here are the steps I take (though note, I'm doing this with Craft 5's bulk edit feature; the plugin may be different):

1. Add the new field to the Entry type
2. Go to the view for that entry in the Craft Admin CP (you _may_ need to create a custom view here: Entries -> left sidebar, bottom three-dots button -> edit views and create a new source)
3. Use the View dropdown to customise the entry view (top right) so that only the old field and new field are visible; it's easiest if the old field comes before the new one, so that you can select the value then traverse across the DOM using JavaScript.
4. Use the search filters (right hand icon button in the search bar) to filter out any unwanted entries. This is also where I'd set a filter to remove any entries that have values on the **new** field, just to make progress easier to track.
5. Use the bulk edit. Both fields should now become editable.
6. Open your developer tools in the browser (typically right-click anywhere on the page and select Inspect, then open the Console panel)
7. Paste a `forEach` function like the one in my original post into the console. Edit anything that needs to be edited e.g. `data-title` values, DOM hierarchy etc. If you're unclear how that code is working, this may not be the best option, but if you know JavaScript it should be okay.
8. Once the console function is ready, hit Enter to execute it. You should see the values cloned from one field to the other (I wouldn't wipe the original value, just as a backup). Save the edited entries, and repeat the process until all entries are edited. You can use the up arrow in the browser console to bring the previous function back, so you don't have to keep copy/pasting it around
9. Update any APIs, templates, etc. to use the new field. Then you can remove the old field.

Hopefully that all makes sense. If you're not used to editing web pages from within the browser itself, this may seem quite complicated, and you may want to familiarise yourself with how the browser console is working by playing around with `querySelector` functions first. You'll certainly need to use the developer tools to understand the DOM structure for your JavaScript to walk along 😉

**Reply by snagelia** [reply]
Thank you. This helps. Yes, I am not used to of editing web pages from within the browser. Will give it a try and update you the outcome.

**Comment by engram-design** [comment]
I know you might not be keen on a paid plugin compared to Typed Link, which was free, but [Hyper](https://plugins.craftcms.com/hyper) has a migration to convert everything over for you, including field content. The other caveat being that this is currently only possible on Craft 4, so it might be too late in the process for your project.

If you don't want to go with Hyper, and want to keep it native with the URL field, take a look at the [migrations](https://github.com/verbb/hyper/tree/craft-4/src/migrations) Hyper has, which are split into two parts - the field and the content.
